tools/cleanup/cleanup.go

```diff
@@ -0,0 +1,308 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// The cleanup tool deletes old GCR images and test clusters from test
+// projects.
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/google"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/pkg/errors"
+	"google.golang.org/api/option"
+	"knative.dev/test-infra/shared/common"
+
+	"knative.dev/pkg/test/gke"
+)
+
+// resourceDeleter is a function that deletes a particular type of resource, and return how many resources were deleted.
+type resourceDeleter func(string, time.Time) (int, error)
```

Comment:
feel like can be a interface

Reply:
Handled this in the comment above about using an OOP style.